File: kv_cache_manager/py_connector/common/static_service_discovery.py
# -*- coding: utf-8 -*-
"""Static Service Discovery：固定 ip:port 列表的本地实现。

适用场景：
    - 测试 / 单机部署直接写死 endpoint 列表
    - 不需要动态订阅但又想统一走 ServiceDiscovery 抽象的下游

URL 形式（由工厂解析后注入）::

    static://11.22.33.44:8080,33.55.66.77:8080
"""

import threading
from typing import List, Optional, Sequence

from kv_cache_manager.py_connector.common.logger import logger
from kv_cache_manager.py_connector.common.service_discovery import (
    ServiceDiscovery,
    ServiceEndpoint,
)


def parse_host_port_list(host_list: str) -> List[ServiceEndpoint]:
    """把 "ip1:port1,ip2:port2" 解析成 ServiceEndpoint 列表。

    任一段非法（缺 port / port 非数 / 空 host）即抛 ValueError。
    """
    if not host_list:
        raise ValueError("host_list is empty")
    endpoints: List[ServiceEndpoint] = []
    for token in host_list.split(','):
        token = token.strip()
        if not token:
            continue
        if ':' not in token:
            raise ValueError(f"static endpoint missing host:port format, token={token!r}")
        host, _, port_str = token.partition(':')
        if not host or not port_str:
            raise ValueError(f"static endpoint missing host or port, token={token!r}")
        if not port_str.isdigit():
            raise ValueError(f"static endpoint port not numeric, token={token!r}")
        port = int(port_str)
        if port <= 0 or port > 65535:
            raise ValueError(f"static endpoint port out of range, token={token!r}")
        endpoints.append(ServiceEndpoint(ip=host, port=port, host=f"{host}:{port}"))
    if not endpoints:
        raise ValueError(f"static endpoint list is empty after parsing, raw={host_list!r}")
    return endpoints


class StaticServiceDiscovery(ServiceDiscovery):
    """静态 ip:port 列表的 ServiceDiscovery 实现。

    Args:
        host_list: "ip1:port1,ip2:port2,..." 形式的字符串
        endpoints: 已解析好的端点列表；与 host_list 二选一
    """

    def __init__(
        self,
        host_list: Optional[str] = None,
        *,
        endpoints: Optional[Sequence[ServiceEndpoint]] = None,
    ):
        if endpoints is not None:
            parsed: List[ServiceEndpoint] = list(endpoints)
        elif host_list is not None:
            parsed = parse_host_port_list(host_list)
        else:
            raise ValueError("must provide either host_list or endpoints")
        if not parsed:
            raise ValueError("StaticServiceDiscovery init with empty endpoints")
        self._endpoints: List[ServiceEndpoint] = parsed
        self._rr_index = 0
        self._rr_lock = threading.Lock()

    def get_type(self) -> str:
        return "Static"

    def get_all_endpoints(self) -> List[ServiceEndpoint]:
        return list(self._endpoints)

    def get_one_endpoint(self) -> Optional[ServiceEndpoint]:
        if not self._endpoints:
            return None
        with self._rr_lock:
            ep = self._endpoints[self._rr_index % len(self._endpoints)]
            self._rr_index = (self._rr_index + 1) % len(self._endpoints)
        return ep

    def refresh(self) -> bool:
        return bool(self._endpoints)

    def close(self) -> None:
        # 无外部资源
        return None

File: kv_cache_manager/common/static_service_discovery.cc
#include "kv_cache_manager/common/static_service_discovery.h"

#include <cctype>
#include <cstdlib>

#include "kv_cache_manager/common/logger.h"

namespace kv_cache_manager {

namespace {

bool ParseSinglePort(const std::string &port_str, int &port_out) {
    if (port_str.empty()) {
        return false;
    }
    for (char c : port_str) {
        if (!std::isdigit(static_cast<unsigned char>(c))) {
            return false;
        }
    }
    long port = std::strtol(port_str.c_str(), nullptr, 10);
    if (port <= 0 || port > 65535) {
        return false;
    }
    port_out = static_cast<int>(port);
    return true;
}

} // namespace

bool StaticServiceDiscovery::ParseHostPortList(const std::string &host_list, std::vector<ServiceEndpoint> &out) {
    out.clear();
    if (host_list.empty()) {
        return false;
    }
    size_t start = 0;
    while (start <= host_list.size()) {
        size_t end = host_list.find(',', start);
        if (end == std::string::npos) {
            end = host_list.size();
        }
        if (end == start) {
            // 连续逗号或末尾逗号 → 跳过空段
            start = end + 1;
            continue;
        }
        const std::string token = host_list.substr(start, end - start);
        const size_t colon = token.find(':');
        if (colon == std::string::npos || colon == 0 || colon == token.size() - 1) {
            KVCM_LOG_ERROR("Static endpoint missing host:port format, token=[%s]", token.c_str());
            out.clear();
            return false;
        }
        int port = 0;
        if (!ParseSinglePort(token.substr(colon + 1), port)) {
            KVCM_LOG_ERROR("Static endpoint port invalid, token=[%s]", token.c_str());
            out.clear();
            return false;
        }
        out.emplace_back(token.substr(0, colon), port);
        start = end + 1;
    }
    if (out.empty()) {
        KVCM_LOG_ERROR("Static endpoint list is empty after parsing, raw=[%s]", host_list.c_str());
        return false;
    }
    return true;
}

bool StaticServiceDiscovery::Init(const std::string &service_address) {
    std::vector<ServiceEndpoint> parsed;
    if (!ParseHostPortList(service_address, parsed)) {
        return false;
    }
    return InitWithEndpoints(std::move(parsed));
}

bool StaticServiceDiscovery::InitWithEndpoints(std::vector<ServiceEndpoint> endpoints) {
    if (endpoints.empty()) {
        KVCM_LOG_ERROR("StaticServiceDiscovery init with empty endpoints");
        return false;
    }
    endpoints_ = std::move(endpoints);
    rr_counter_.store(0, std::memory_order_relaxed);
    return true;
}

bool StaticServiceDiscovery::GetAllEndpoints(std::vector<ServiceEndpoint> &endpoints) {
    if (endpoints_.empty()) {
        endpoints.clear();
        return false;
    }
    endpoints = endpoints_;
    return true;
}

bool StaticServiceDiscovery::GetOneEndpoint(ServiceEndpoint &endpoint) {
    if (endpoints_.empty()) {
        return false;
    }
    const size_t idx = rr_counter_.fetch_add(1, std::memory_order_relaxed) % endpoints_.size();
    endpoint = endpoints_[idx];
    return true;
}

} // namespace kv_cache_manager
